presto-main/src/main/java/com/facebook/presto/sql/gen/TryCodeGenerator.java

```diff
@@ -0,0 +1,139 @@
+{
+    private static final List<StandardErrorCode> CAUGHT_EXCEPTION_CODES = ImmutableList.of(DIVISION_BY_ZERO, INVALID_CAST_ARGUMENT, INVALID_FUNCTION_ARGUMENT, NUMERIC_VALUE_OUT_OF_RANGE);
+    private static final String EXCEPTION_HANDLER_NAME = "catchExceptionMethod";
+
+    private final CallExpression call;
+    private final Map<CallExpression, MethodDefinition> tryMethodsMap;
+    private final List<Variable> inputParameters;
+
+    public TryCodeGenerator(CallExpression call, Map<CallExpression, MethodDefinition> tryMethodsMap, List<Variable> inputParameters)
+    {
+        this.call = call;
+        this.tryMethodsMap = tryMethodsMap;
+        this.inputParameters = inputParameters;
+    }
+
+    @Override
+    public BytecodeNode generateExpression(Signature signature, BytecodeGeneratorContext context, Type returnType, List<RowExpression> arguments)
+    {
+        checkState(tryMethodsMap.containsKey(call), "try methods map does not contain this try call");
+
+        BytecodeBlock visitorMethodBody = new BytecodeBlock()
+                .comment("load required variables")
+                .getVariable(context.getScope().getVariable("this"));
+
+        inputParameters.stream()
+                .forEach(visitorMethodBody::getVariable);
+
+        visitorMethodBody.comment("call dynamic try method: " + tryMethodsMap.get(call).getName())
+                .invokeVirtual(tryMethodsMap.get(call))
+                .append(unboxPrimitiveIfNecessary(context.getScope(), wrap(call.getType().getJavaType())));
+
+        return visitorMethodBody;
+    }
+
+    public static MethodDefinition defineTryMethod(
+            ClassDefinition classDefinition,
+            BytecodeExpressionVisitor innerExpressionVisitor,
+            RowExpression innerRowExpression,
+            List<Parameter> inputParameters, Class<?> returnType,
+            CallSiteBinder callSiteBinder,
+            CachedInstanceBinder cachedInstanceBinder,
+            FunctionRegistry functionRegistry)
+    {
+        MethodDefinition method = classDefinition.declareMethod(a(PUBLIC), "try_" + Math.abs(innerRowExpression.hashCode()), type(returnType), inputParameters);
+        Scope calleeMethodScope = method.getScope();
+
+        BytecodeGeneratorContext calleeMethodContext = new BytecodeGeneratorContext(innerExpressionVisitor, calleeMethodScope, callSiteBinder, cachedInstanceBinder, functionRegistry);
+        BytecodeNode innerExpression = calleeMethodContext.generate(innerRowExpression);
+
+        MethodType exceptionHandlerType = methodType(returnType, RuntimeException.class);
+        MethodHandle exceptionHandler = methodHandle(TryCodeGenerator.class, EXCEPTION_HANDLER_NAME, RuntimeException.class).asType(exceptionHandlerType);
+        Binding binding = callSiteBinder.bind(exceptionHandler);
+
+        method.comment("Try projection: %s", innerRowExpression.toString());
+        method.getBody()
+                .append(new TryCatch(
+                        "",
```

Comment:
The first argument is allowed to be null. I would add a constructor to TryCatch that doesn't take the comment argument.